src/dfx/src/commands/canister/call.rs

```diff
@@ -136,91 +135,71 @@ async fn request_id_via_wallet_call(
-                Err(DiagnosedError::new(
-                    format!(
-                        "{} can only be called by a canister, not by an external user.",
-                        method_name.as_ref()
-                    ),
-                    format!("The easiest way to call {} externally is to proxy this call through a wallet. Try calling this with 'dfx canister call <other arguments> (--network ic) --wallet <wallet id>'.\n\
-                    To figure out the id of your wallet, run 'dfx identity get-wallet (--network ic)'.", method_name.as_ref())
-                )).context("Method only callable by a canister.")
-            }
-            MgmtMethod::InstallCode => {
-                let install_args = candid::Decode!(arg_value, CanisterInstall)
-                    .context("Failed to decode arguments.")?;
-                Ok(install_args.canister_id)
-            }
-            MgmtMethod::UpdateSettings => {
-                #[derive(CandidType, Deserialize)]
-                struct In {
-                    canister_id: CanisterId,
-                    settings: CanisterSettings,
-                }
-                let in_args =
-                    candid::Decode!(arg_value, In).context("Failed to decode arguments.")?;
-                Ok(in_args.canister_id)
-            }
-            MgmtMethod::StartCanister
-            | MgmtMethod::StopCanister
-            | MgmtMethod::CanisterStatus
-            | MgmtMethod::DeleteCanister
-            | MgmtMethod::DepositCycles
-            | MgmtMethod::UninstallCode
-            | MgmtMethod::ProvisionalTopUpCanister
-            | MgmtMethod::UploadChunk
-            | MgmtMethod::ClearChunkStore
-            | MgmtMethod::StoredChunks
-            | MgmtMethod::FetchCanisterLogs => {
-                #[derive(CandidType, Deserialize)]
-                struct In {
-                    canister_id: CanisterId,
-                }
-                let in_args =
-                    candid::Decode!(arg_value, In).context("Failed to decode arguments.")?;
-                Ok(in_args.canister_id)
-            }
-            MgmtMethod::ProvisionalCreateCanisterWithCycles => {
-                Ok(CanisterId::management_canister())
+    match method_name {
+        MgmtMethod::CreateCanister
+        | MgmtMethod::RawRand
+        | MgmtMethod::BitcoinGetBalance
+        | MgmtMethod::BitcoinGetUtxos
+        | MgmtMethod::BitcoinSendTransaction
+        | MgmtMethod::BitcoinGetCurrentFeePercentiles
+        | MgmtMethod::EcdsaPublicKey
+        | MgmtMethod::SignWithEcdsa
+        | MgmtMethod::NodeMetricsHistory => Ok(CanisterId::management_canister()),
+        MgmtMethod::InstallCode => {
+            // TODO: Maybe this case can be merged with the following one.
+            let install_args = candid::Decode!(arg_value, CanisterInstall)
+                .context("Failed to decode arguments.")?;
+            Ok(install_args.canister_id)
```

Comment:
Do IIUC that we now allow CreateCanister, RawRand and all these other methods in ingress messages by external users?

The PR description only talks about changes that allow MgmtMethod::BitcoinGetBalanceQuery | MgmtMethod::BitcoinGetUtxosQuery to bypass query signatures verification (line 275 of call.rs).

Reply:
I moved the logic of "Is this call only allowed from canisters?" from this `get_effective_canister_id()` method to the following `exec()` method.
https://github.com/dfinity/sdk/blob/38b34351867dd44b011731980ab6db16414ceab0/src/dfx/src/commands/canister/call.rs#L298-L325